File: chapter02/arrayqueue.go
package ods

import "math"

// Adheres to FIFO queue.
type ArrayQueue struct {
	a []interface{}
	j int
	n int
}

func (a *ArrayQueue) Size() int {
	return a.n
}

func (a *ArrayQueue) Init(i int) {
	a.a = make([]interface{}, i)
	a.j = 0
	a.n = 0
}

func (a *ArrayQueue) Add(i int, x interface{}) bool {
	if a.n+1 >= len(a.a) {
		a.resize()
	}
	a.a[(a.j+a.n)%len(a.a)] = x
	a.n++
	return true
}

func (a *ArrayQueue) Remove() interface{} {
	x := a.a[a.j]
	a.j = (a.j + 1) % len(a.a)
	a.n--
	if len(a.a) >= 3*a.n {
		a.resize()
	}
	return x
}

func (a *ArrayQueue) resize() {
	b := make([]interface{}, int(math.Max(1, float64(2*a.n))))
	for k := 0; k < a.n; k++ {
		b[k] = a.a[(a.j+k)%len(a.a)]
	}
	a.a = b
	a.j = 0
}

File: chapter02/src/arrayqueue.rs
use chapter01::interface::Queue;

#[derive(Clone, Debug, Default)]
pub struct Array<T> {
    a: Box<[Option<T>]>,
    j: usize,
    n: usize,
}

impl<T: Clone> Array<T> {
    pub fn new() -> Self {
        Self::with_length(1)
    }

    pub fn length(&self) -> usize {
        self.a.len()
    }

    fn allocate_on_heap(size: usize) -> Box<[Option<T>]> {
        vec![Default::default(); size].into_boxed_slice()
    }

    pub fn with_length(capacity: usize) -> Self {
        Self {
            a: Self::allocate_on_heap(capacity),
            j: 0,
            n: 0,
        }
    }

    fn resize(&mut self) {
        let mut b = Self::allocate_on_heap(std::cmp::max(self.n * 2, 1));
        std::mem::swap(&mut self.a, &mut b);
        for i in 0..self.n {
            self.a[i] = b[(self.j + i) % b.len()].take();
        }
        self.j = 0;
    }
}

impl<T: Clone> Queue<T> for Array<T> {
    fn add(&mut self, x: T) -> () {
        if self.n + 1 >= self.length() {
            self.resize();
        }
        self.a[(self.j + self.n) % self.length()] = Some(x);
        self.n += 1;
    }

    fn remove(&mut self) -> Option<T> {
        let x = self.a[self.j].take();
        self.j = (self.j + 1) % self.length();
        self.n -= 1;
        if self.length() >= 3 * self.n {
            self.resize();
        }
        x
    }
}

#[cfg(test)]
mod test {
    use super::Array;
    use chapter01::interface::Queue;
    #[test]
    fn test_arrayqueue() {
        let mut array_queue: Array<char> = Array::new();
        for elem in "aaabc".chars() {
            array_queue.add(elem);
        }
        assert_eq!(array_queue.remove(), Some('a'));
        assert_eq!(array_queue.remove(), Some('a'));
        array_queue.add('d');
        array_queue.add('e');
        assert_eq!(array_queue.remove(), Some('a'));
        array_queue.add('f');
        array_queue.add('g');
        assert_eq!(array_queue.length(), 10);
        array_queue.add('h');
        assert_eq!(array_queue.remove(), Some('b'));
        println!("\nArrayQueue = {:?}\n", array_queue);
    }
}
